### PR TITLE
Client Side Caching

### PR DESCRIPTION
**Pull request by altanozlu**
Hello everyone, 
Client Side Caching is almost ready, just needs more command coverage and more testing.
I changed creating of new clients with ConnectionConfig because I think it will be ugly to repeating with_cache functions everywhere.
I had to change pin-project-lite to pin-project since lite pin-project-lite doesn't support cfg feature in it.

### CODE REVIEW

**Comment by nihohit** [comment]
This is a very large PR, containing multiple layers of changes, which should be split into separate PRs in order to allow for better review.
1. the changes to the `Client` API aren't backwards compatible - at the very least we should deprecate the old functions with a message pointing towards the right function, instead of just removing the old functions. this should be a leading PR, and the Config object should use a builder pattern, similar to what's used in the cluster client's config.
2. can you explain the changes to `pin-project-lite`? where in this PR can we see that requirement? In any case, this, too, should be a separate PR.
3. I haven't gone over the whole code yet. If there are other sub-changes, they should be spun out to separate PRs too.

**Comment by nihohit** [comment]
related to using a config object - https://github.com/redis-rs/redis-rs/issues/1147#issuecomment-2047893934
not sure why I thought that the config needs the builder pattern. It's nice, but not required.

**Comment by altanozlu** [comment]
> > By being like Cluster client's config you mean ConnectionConfigBuilder.build should create the connection ? Or it shouldn't contain ConnectionConfig.
> 
> Sorry, I don't know why I didn't answer this beforehand. I think that if you'll take `Client::get_connection_manager_with_backoff_and_timeouts`, and the other various `Client::get_connection_manager*` functions, they should become something like
> 
> ```rust
> client.get_tokio_connection_manager_with_config(ConnectionManagerConfig {
>   response_timeout: Duration::from_millis(10), 
>   Default::default()
> })
> ```
> 
> or
> 
> ```rust
> client.get_tokio_connection_manager_with_config( 
>   ConnectionManagerBuilder::new()
>     .set_response_timeout(Duration::from_millis(10))
>     .build() 
> })
> ```
> 
> completely removing `Client`, or just leaving it for backwards compatibility, is also valid, but might be too much of a breaking change. We'll need to update all of the documentation, for starters.
> 
> ```rust
> ConnectionManagerBuilder::new(address)
>    .set_authentication(auth)
>    .set_response_timeout(Duration::from_millis(10))
>    .connect().await
> ```

I also like the first approach, I'll wait for main branch to get this approach and rebase over it then.

**Comment by nihohit** [comment]
a configuration object was added for ConnectionManager in this PR:
https://github.com/redis-rs/redis-rs/pull/1194

**Comment by altanozlu** [comment]
> a configuration object was added for ConnectionManager in this PR: #1194

I will be on vacation for 3 weeks then I will work on this PR again.

**Comment by altanozlu** [comment]
I've updated the branch and pin-project and parking_lot is not used anymore, I will post benchmark results soon.

**Comment by nihohit** [comment]
Hi, I'm sorry, but I couldn't find the time to review this yet. I see that this is still a draft. Are there specific areas / design choices you particularly feel that require guidance or help?
Once this will be ready for review, could you *save this branch*, and open another PR, with just a cache for the multiplexed connection? A smaller PR could get in much faster.

File: redis/Cargo.toml [hunk 1]
```diff
@@ -119,6 +122,7 @@ bigdecimal = ["dep:bigdecimal"]
 num-bigint = []
 uuid = ["dep:uuid"]
 disable-client-setinfo = []
+cache = ["aio", "lru", "once_cell"]
```
[the remark this answers is not in the document]
Comment (altanozlu) [reply redis/Cargo.toml:125]: I was planning to make it only usable on multiplexed connection, to make it future proof ofc cache_aio makes more sense 👌🏻

**Comment by altanozlu** [comment]
> Hi, I'm sorry, but I couldn't find the time to review this yet. I see that this is still a draft. Are there specific areas / design choices you particularly feel that require guidance or help? Once this will be ready for review, could you _save this branch_, and open another PR, with just a cache for the multiplexed connection? A smaller PR could get in much faster.

Hi! I kinda wait for myself to prepare good benchmark results and evaluating results. I will open new PR with only MPX next weekend.

**Comment by nihohit** [comment]
thanks! no need to close this, I just wanted a smaller scope for the initial review.

**Pull request closed without merging**